### PR TITLE
Postgres: avoid unnecessary casts of DATE or TIMESTAMPTZ to TIMESTAMP

### LINKED ISSUE #19816: Postgres unnecessarily casts `timestamptz` and `date` columns to `timestamp` inside `date_trunc()`
In #11502 I added logic to avoid `cast(field AS 'date')` or `field::date` in Postgres if `field` was already a `DATE`. Similarly, it's smart enough to avoid casts to `timestamp` if the `field` is already a `timestamp`.

However, it still casts to `timestamp` if the field is a `timestamptz` (`timestamp with time zone`) or a `date`, altho this cast is completely unnecessary.  Avoiding these casts is important since in some cases it can make a pretty big performance difference in queries 

### EXAMPLE QUERY

```sql
SELECT 
  "Q2"."created_at" AS "Q2__created_at",
  "source"."count"  AS "count",
  "Q2"."birth_date" AS "Q2__birth_date",
  "Q2"."count"      AS "Q2__count" 
FROM (
  SELECT 
    date_trunc('month', CAST("public"."people"."created_at" AS timestamp)) AS "created_at",
    count(*)                                                               AS "count" 
  FROM 
    "public"."people" 
  GROUP BY 
    date_trunc('month', CAST("public"."people"."created_at" AS timestamp)) 
  ORDER BY 
    date_trunc('month', CAST("public"."people"."created_at" AS timestamp)) ASC
) "source" 
LEFT JOIN (
  SELECT 
    date_trunc('month', CAST("public"."people"."birth_date" AS timestamp)) AS "birth_date",
    count(*)                                                               AS "count" 
  FROM 
    "public"."people" 
  GROUP BY 
    date_trunc('month', CAST("public"."people"."birth_date" AS timestamp)) 
  ORDER BY 
    date_trunc('month', CAST("public"."people"."birth_date" AS timestamp)) ASC
) "Q2"
  ON date_trunc('month', CAST("source"."created_at" AS timestamp))
   = date_trunc('month', CAST("Q2"."birth_date" AS timestamp)) 
LIMIT 3
```

### WHAT IT SHOULD LOOK LIKE

```sql
SELECT 
  "Q2"."created_at" AS "Q2__created_at",
  "source"."count"  AS "count",
  "Q2"."birth_date" AS "Q2__birth_date",
  "Q2"."count"      AS "Q2__count" 
FROM (
  SELECT 
    date_trunc('month', "public"."people"."created_at") AS "created_at",
    count(*)                                            AS "count" 
  FROM 
    "public"."people" 
  GROUP BY 
    date_trunc('month', "public"."people"."created_at") 
  ORDER BY 
    date_trunc('month', "public"."people"."created_at") ASC
) "source" 
LEFT JOIN (
  SELECT 
    date_trunc('month', "public"."people"."birth_date") AS "birth_date",
    count(*)                                            AS "count" 
  FROM 
    "public"."people" 
  GROUP BY 
    date_trunc('month', "public"."people"."birth_date") 
  ORDER BY 
    date_trunc('month', "public"."people"."birth_date") ASC
) "Q2"
  ON date_trunc('month', "source"."created_at")
   = date_trunc('month', "Q2"."birth_date") 
LIMIT 3
```

### PR DESCRIPTION
**Pull request by camsaul**
Fixes #19816

I noticed this while working on #18502 thanks to the new `with-native-query-testing-context` and `sql=` stuff I added in #19384.

In #11502 I added logic to avoid `cast(field AS 'date')` or `field::date` in Postgres if `field` was already a `DATE`. Similarly, it's smart enough to avoid casts to `timestamp` if the `field` is already a `timestamp`.

However, it still casts to `timestamp` if the field is a `timestamptz` (`timestamp with time zone`) or a `date`, altho this cast is completely unnecessary. 

This PR adds a bit of logic to check whether a column is of some set of allowed types and only casts if it is not one of those types. Example query:

### BEFORE

```sql
SELECT 
  "Q2"."created_at" AS "Q2__created_at",
  "source"."count"  AS "count",
  "Q2"."birth_date" AS "Q2__birth_date",
  "Q2"."count"      AS "Q2__count" 
FROM (
  SELECT 
    date_trunc('month', CAST("public"."people"."created_at" AS timestamp)) AS "created_at",
    count(*)                                                               AS "count" 
  FROM 
    "public"."people" 
  GROUP BY 
    date_trunc('month', CAST("public"."people"."created_at" AS timestamp)) 
  ORDER BY 
    date_trunc('month', CAST("public"."people"."created_at" AS timestamp)) ASC
) "source" 
LEFT JOIN (
  SELECT 
    date_trunc('month', CAST("public"."people"."birth_date" AS timestamp)) AS "birth_date",
    count(*)                                                               AS "count" 
  FROM 
    "public"."people" 
  GROUP BY 
    date_trunc('month', CAST("public"."people"."birth_date" AS timestamp)) 
  ORDER BY 
    date_trunc('month', CAST("public"."people"."birth_date" AS timestamp)) ASC
) "Q2"
  ON date_trunc('month', CAST("source"."created_at" AS timestamp))
   = date_trunc('month', CAST("Q2"."birth_date" AS timestamp)) 
LIMIT 3
```

### AFTER 

```sql
SELECT 
  "Q2"."created_at" AS "Q2__created_at",
  "source"."count"  AS "count",
  "Q2"."birth_date" AS "Q2__birth_date",
  "Q2"."count"      AS "Q2__count" 
FROM (
  SELECT 
    date_trunc('month', "public"."people"."created_at") AS "created_at",
    count(*)                                            AS "count" 
  FROM 
    "public"."people" 
  GROUP BY 
    date_trunc('month', "public"."people"."created_at") 
  ORDER BY 
    date_trunc('month', "public"."people"."created_at") ASC
) "source" 
LEFT JOIN (
  SELECT 
    date_trunc('month', "public"."people"."birth_date") AS "birth_date",
    count(*)                                            AS "count" 
  FROM 
    "public"."people" 
  GROUP BY 
    date_trunc('month', "public"."people"."birth_date") 
  ORDER BY 
    date_trunc('month', "public"."people"."birth_date") ASC
) "Q2"
  ON date_trunc('month', "source"."created_at")
   = date_trunc('month', "Q2"."birth_date") 
LIMIT 3
```

### CODE REVIEW

File: src/metabase/util/honeysql_extensions.clj [hunk 1]
```diff
@@ -257,6 +257,17 @@
       expr
       (cast sql-type expr)))
 
+(defn cast-unless-type-in
```
Comment (jeff303) [inline src/metabase/util/honeysql_extensions.clj:260]: Add `:added` metadata since new public symbol?

**Review by jeff303 — APPROVED** [review]

File: src/metabase/util/honeysql_extensions.clj (hunk 1 above)
Comment (camsaul) [reply src/metabase/util/honeysql_extensions.clj:260]: probably a good idea. I'll add it

File: src/metabase/util/honeysql_extensions.clj [hunk 2]
```diff
@@ -257,6 +257,17 @@
...
+  "Cast `expr` to `desired-type` unless `expr` is of one of the `acceptable-types`. Returns a typed HoneySQL form.
+
+    ;; cast to TIMESTAMP unless form is already a TIMESTAMP, TIMESTAMPTZ, or DATE
+    (cast-unless-type-in \"timestamp\" #{\"timestamp\" \"timestamptz\" \"date\"} form)"
+  [desired-type acceptable-types expr]
```
Comment (camsaul) [inline src/metabase/util/honeysql_extensions.clj:265]:
```suggestion
  {:added "0.42.0"}
  [desired-type acceptable-types expr]
```

File: src/metabase/util/honeysql_extensions.clj (hunk 1 above)
Comment (camsaul) [reply src/metabase/util/honeysql_extensions.clj:260]: added

File: test/metabase/driver/postgres_test.clj [hunk 3]
```diff
@@ -665,6 +665,23 @@
                       "ORDER BY attempts.date ASC")
                  (some-> (qp/query->native query) :query pretty-sql))))))))
 
+(deftest do-not-cast-to-timestamp-if-column-if-timestamp-tz-or-date-test
+  (testing "Don't cast a DATE or TIMESTAMPTZ to TIMESTAMP, it's not necessary"
```
Comment (camsaul) [inline test/metabase/driver/postgres_test.clj:669]:
```suggestion
  (testing "Don't cast a DATE or TIMESTAMPTZ to TIMESTAMP, it's not necessary (#19816)"
```

**Comment by rlotun** [comment]
do you think this should be backported to `release-x.42.x`? it isn't there now

**Pull request merged**